### PR TITLE
BACK - Routes cryptos

### PR DESCRIPTION
**Pull request by Gerroll**
**What does this PR do?**

This PR add all the routes require in the spec for crypto currency

**How should this be manually tested?**
You must follow the pull request #10 first because changes were made.
Then please follow the next commands :
- cd backend
- npm i
- npm start
- Try routes define in file backend/routes/cryptos.js (there are some example in it)

**Other changes**
#10 => field 'period' now match specification by taking the one of the following values : 'daily', 'hourly', 'minute' 
/
**Boyscout**
In the futur, swagger should be implemented
/

### CODE REVIEW

File: backend/routes/cryptos.js [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    - /cryptos?cmids=BTC,ETH,XRP
+*/
+router.get('/cryptos', function (req, res, next) {
+    let query1 = "SELECT * from CRYPTO_LIST",
+        values = "";
+    if (req.query.cmids != undefined) {
+        const cmids = req.query.cmids.split(",");
+        values = `'${cmids[0]}'`
+        for (let i = 1; i < cmids.length; i++)
+            values += `,'${cmids[i]}'`
```
Comment (berthieresteban) [inline backend/routes/cryptos.js]: You can use here some JS arrays methods :
```values = req.query.cmids.split(",").map(cmid => `'${cmid}'`).toString();```

File: backend/routes/cryptos.js [hunk 2]
```diff
@@ -0,0 +1,204 @@
+var express = require('express');
+var router = express.Router();
```
Comment (Yoann-Abbes) [inline backend/routes/cryptos.js]: Don't use `var`

File: backend/app.js [hunk 3]
```diff
@@ -3,9 +3,9 @@ var path = require('path');
 var cookieParser = require('cookie-parser');
 var logger = require('morgan');
 
-var indexRouter = require('./routes/index');
-var userRouter = require('./routes/user');
-var adminRouter = require('./routes/admin');
+var cryptoRouter = require('./routes/cryptos');
```
Comment (berthieresteban) [inline backend/app.js]:
```suggestion
const cryptoRouter = require('./routes/cryptos');
```

File: backend/app.js [hunk 4]
```diff
@@ -3,9 +3,9 @@ var path = require('path');
...
+var userRouter = require('./routes/users');
```
Comment (berthieresteban) [inline backend/app.js]:
```suggestion
const userRouter = require('./routes/users');
```

File: backend/app.js [hunk 5]
```diff
@@ -3,9 +3,9 @@ var path = require('path');
...
+var articlesRouter = require('./routes/articles');
```
Comment (berthieresteban) [inline backend/app.js]:
```suggestion
const articlesRouter = require('./routes/articles');
```

File: backend/routes/cryptos.js [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+/* GET cryptos infos
+/cryptos?[cmids={cm}]
+examples:
+    - /cryptos
+    - /cryptos?cmids=BTC
+    - /cryptos?cmids=BTC,ETH,XRP
+*/
+router.get('/cryptos', function (req, res, next) {
+    let query1 = "SELECT * from CRYPTO_LIST",
```
Comment (Yoann-Abbes) [inline backend/routes/cryptos.js]:
```suggestion
    const GET_CRYPTO_LIST = "SELECT * from CRYPTO_LIST",
```
Also, should be a global

File: backend/routes/cryptos.js [hunk 7]
```diff
@@ -0,0 +1,204 @@
+var express = require('express');
```
Comment (berthieresteban) [inline backend/routes/cryptos.js]:
```suggestion
const express = require('express');
```

File: backend/routes/cryptos.js (hunk 2 above)
Comment (berthieresteban) [inline backend/routes/cryptos.js]:
```suggestion
const router = express.Router();
```

File: backend/routes/cryptos.js [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+/cryptos?[cmids={cm}]
+examples:
+    - /cryptos
+    - /cryptos?cmids=BTC
+    - /cryptos?cmids=BTC,ETH,XRP
+*/
+router.get('/cryptos', function (req, res, next) {
+    let query1 = "SELECT * from CRYPTO_LIST",
+        values = "";
+    if (req.query.cmids != undefined) {
```
Comment (Yoann-Abbes) [inline backend/routes/cryptos.js]:
```suggestion
    if (req.query && req.query.cmids) {
```

Comment (berthieresteban) [inline backend/routes/cryptos.js]:
```suggestion
    if (req.query.cmids !== undefined) {
```

File: backend/routes/cryptos.js [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+        const cmids = req.query.cmids.split(",");
+        values = `'${cmids[0]}'`
+        for (let i = 1; i < cmids.length; i++)
+            values += `,'${cmids[i]}'`
+        query1 = `SELECT id,symbol,fullname,picture_url from CRYPTO_LIST WHERE symbol IN (${values})`
```
Comment (Yoann-Abbes) [inline backend/routes/cryptos.js]:
```suggestion
        const GET_CRYPTO_LIST_BY_SYMBOL = `SELECT id,symbol,fullname,picture_url from CRYPTO_LIST WHERE symbol IN (${values})`
```
Should be global

File: backend/routes/cryptos.js [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+    }
+    client
+        .query(query1)
```
Comment (Yoann-Abbes) [inline backend/routes/cryptos.js]:
```suggestion
        .query(GET_CRYPTO_LIST_BY_SYMBOL)
```

File: backend/routes/cryptos.js [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+        .then(result1 => {
```
Comment (Yoann-Abbes) [inline backend/routes/cryptos.js]:
```suggestion
        .then(result => {
```

File: backend/routes/cryptos.js [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+            let now = new Date()
+            let a = new Date(now.getFullYear(), now.getMonth(), now.getDate(), 1, 0, 0).getTime() / 1000
+            if (result1.rows.length == 0)
```
Comment (Yoann-Abbes) [inline backend/routes/cryptos.js]:
```suggestion
            if (result.rows.length == 0)
```

File: backend/routes/cryptos.js [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+        const cmids = req.query.cmids.split(",");
+        values = `'${cmids[0]}'`
+        for (let i = 1; i < cmids.length; i++)
+            values += `,'${cmids[i]}'`
+        query1 = `SELECT id,symbol,fullname,picture_url from CRYPTO_LIST WHERE symbol IN (${values})`
+    }
+    client
+        .query(query1)
+        .then(result1 => {
+            let now = new Date()
```
Comment (Yoann-Abbes) [inline backend/routes/cryptos.js]:
```suggestion
            const now = new Date()
```

File: backend/routes/cryptos.js (hunk 12 above)
Comment (berthieresteban) [inline backend/routes/cryptos.js]:
```suggestion
            if (result1.rows.length === 0)
```

File: backend/routes/cryptos.js [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+            let a = new Date(now.getFullYear(), now.getMonth(), now.getDate(), 1, 0, 0).getTime() / 1000
```
Comment (Yoann-Abbes) [inline backend/routes/cryptos.js]: Should be `const` and, have a proper name ^^

File: backend/routes/cryptos.js [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+            if (result1.rows.length == 0)
+                res.status(400).json({
+                    error: `No crypto matches symbols '${cmids}'`
+                })
+            else {
+                let query2 = `SELECT * from CRYPTO_HISTORY WHERE period = 'daily' AND crypto_id IN ( SELECT id FROM CRYPTO_LIST ) AND timestamp = to_timestamp(${a})`
+                if (values != "")
```
Comment (berthieresteban) [inline backend/routes/cryptos.js]:
```suggestion
                if (values !== "")
```

File: backend/routes/cryptos.js [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+*/
+router.get('/cryptos/:cmid', function (req, res, next) {
+    let cmid = req.params.cmid,
+        query1 = `SELECT * from CRYPTO_LIST WHERE symbol = '${cmid}'`;
+    client
+        .query(query1)
+        .then(result1 => {
+            let now = new Date()
+            let a = new Date(now.getFullYear(), now.getMonth(), now.getDate(), 1, 0, 0).getTime() / 1000
+            if (result1.rows.length == 0)
```
Comment (berthieresteban) [inline backend/routes/cryptos.js]:
```suggestion
            if (result1.rows.length === 0)
```

File: backend/routes/cryptos.js [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+router.get('/cryptos/:cmid/history/:period', function (req, res, next) {
+    const cmid = req.params.cmid,
+        period = req.params.period,
+        validPeriod = ['daily', 'hourly', 'minute'];
+    if (validPeriod.includes(period)) {
+        let query = `SELECT period,timestamp,open,high,low,close from CRYPTO_HISTORY WHERE period = '${period}' AND crypto_id = ( SELECT id FROM CRYPTO_LIST WHERE symbol = '${cmid}' LIMIT 1 )`
+        client
+            .query(query)
+            .then(result => {
+                if (result.rows.length == 0)
```
Comment (berthieresteban) [inline backend/routes/cryptos.js]:
```suggestion
                if (result.rows.length === 0)
```

File: backend/routes/cryptos.js [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+            else {
+                let query2 = `SELECT * from CRYPTO_HISTORY WHERE period = 'daily' AND crypto_id IN ( SELECT id FROM CRYPTO_LIST ) AND timestamp = to_timestamp(${a})`
+                if (values != "")
+                    query2 = `SELECT * from CRYPTO_HISTORY WHERE period = 'daily' AND crypto_id IN ( SELECT id FROM CRYPTO_LIST WHERE symbol IN (${values})) AND timestamp = to_timestamp(${a})`
+                client
+                    .query(query2)
+                    .then(result2 => {
+                        result1.rows.forEach(elem1 => {
+                            result2.rows.forEach(elem2 => {
+                                if (elem1.id == elem2.crypto_id) {
```
Comment (berthieresteban) [inline backend/routes/cryptos.js]:
```suggestion
                                if (elem1.id === elem2.crypto_id) {
```

File: backend/routes/cryptos.js [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+        .query(query1)
+        .then(result1 => {
+            let now = new Date()
+            let a = new Date(now.getFullYear(), now.getMonth(), now.getDate(), 1, 0, 0).getTime() / 1000
+            if (result1.rows.length == 0)
+                res.status(400).json({
+                    error: `No crypto matches symbols '${cmids}'`
+                })
+            else {
+                let query2 = `SELECT * from CRYPTO_HISTORY WHERE period = 'daily' AND crypto_id IN ( SELECT id FROM CRYPTO_LIST ) AND timestamp = to_timestamp(${a})`
```
Comment (Yoann-Abbes) [inline backend/routes/cryptos.js]:
```suggestion
                const GET_CRYPTO_HISTORY_BY_DAY_AND_ID = `SELECT * from CRYPTO_HISTORY WHERE period = 'daily' AND crypto_id IN ( SELECT id FROM CRYPTO_LIST ) AND timestamp = to_timestamp(${a})`
```
Also, should be const, and to forget to change `to_timestamp(${a})` with the correct variable name

File: backend/routes/cryptos.js [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+                    error: `No crypto matches symbols '${cmid}'`
+                })
+            else {
+                let query2 = `SELECT * from CRYPTO_HISTORY WHERE period = 'daily' AND crypto_id IN ( SELECT id FROM CRYPTO_LIST WHERE symbol = '${cmid}' ) AND timestamp = to_timestamp(${a})`
+                client
+                    .query(query2)
+                    .then(result2 => {
+                        result1.rows.forEach(elem1 => {
+                            result2.rows.forEach(elem2 => {
+                                if (elem1.id == elem2.crypto_id) {
```
Comment (berthieresteban) [inline backend/routes/cryptos.js]:
```suggestion
                                if (elem1.id === elem2.crypto_id) {
```

File: backend/routes/cryptos.js [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+            res.status(400).end()
+            return
+        }
+    });
+
+    let query1 = `SELECT * FROM CRYPTO_LIST WHERE symbol = '${symbol}'`;
+    client
+        .query(query1)
+        .then(result => {
+            if (result.rows.length == 0) {
```
Comment (berthieresteban) [inline backend/routes/cryptos.js]:
```suggestion
            if (result.rows.length === 0) {
```

File: backend/routes/cryptos.js [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+            let now = new Date()
+            let a = new Date(now.getFullYear(), now.getMonth(), now.getDate(), 1, 0, 0).getTime() / 1000
+            if (result1.rows.length == 0)
+                res.status(400).json({
+                    error: `No crypto matches symbols '${cmids}'`
+                })
+            else {
+                let query2 = `SELECT * from CRYPTO_HISTORY WHERE period = 'daily' AND crypto_id IN ( SELECT id FROM CRYPTO_LIST ) AND timestamp = to_timestamp(${a})`
+                if (values != "")
+                    query2 = `SELECT * from CRYPTO_HISTORY WHERE period = 'daily' AND crypto_id IN ( SELECT id FROM CRYPTO_LIST WHERE symbol IN (${values})) AND timestamp = to_timestamp(${a})`
```
Comment (Yoann-Abbes) [inline backend/routes/cryptos.js]: Same remark than before

File: backend/scripts/db_sample.sql [hunk 23]
```diff
@@ -3,19 +3,17 @@ CREATE DATABASE money;
 \c money;
 DROP TABLE IF EXISTS RSS_LIST CASCADE;
 CREATE TABLE RSS_LIST (
-    id serial PRIMARY KEY,
-    link VARCHAR(500) NOT NULL,
-    description TEXT
+ id serial PRIMARY KEY,
+ link VARCHAR(500) NOT NULL,
+ description TEXT
```
Comment (berthieresteban) [inline backend/scripts/db_sample.sql]: Why ?

File: backend/routes/cryptos.js [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
+                client
+                    .query(query2)
```
Comment (Yoann-Abbes) [inline backend/routes/cryptos.js]: Same remark than before

I stop there for this file

File: backend/scripts/db_sample.sql [hunk 25]
```diff
@@ -30,17 +28,16 @@ CREATE TABLE CRYPTO_HISTORY (
...
-    keyword VARCHAR(500) [],
-    favorites_crypto BIGINT []
+ id serial PRIMARY KEY,
+ is_admin BOOLEAN NOT NULL,
+ email VARCHAR(500) NOT NULL,
+ password VARCHAR(500) NOT NULL,
+ username VARCHAR(500) NOT NULL,
+ picture_url TEXT,
+ keyword VARCHAR(500) [],
+ favorites_crypto BIGINT []
```
Comment (berthieresteban) [inline backend/scripts/db_sample.sql]: Why ?

File: backend/scripts/db_sample.sql [hunk 26]
```diff
@@ -74,23 +71,22 @@ CREATE TABLE USERS (
...
-    ARRAY ['bitcoin', 'etherum'],
-    ARRAY [1, 2]
+ nextval('users_id_seq'),
+ false,
+ 'yoann.abbes@gmail.com',
+ 'testmdp',
+ 'testuser',
+ 'http://osef.com',
+ ARRAY ['bitcoin', 'etherum'],
+ ARRAY [1, 2]
```
Comment (berthieresteban) [inline backend/scripts/db_sample.sql]: Same

File: backend/scripts/db_sample.sql [hunk 27]
```diff
@@ -74,23 +71,22 @@ CREATE TABLE USERS (
...
\ No newline at end of file
+ nextval('users_id_seq'),
+ false,
+ 'yoann1.abbes@epitech.eu',
+ 'testmdp2',
+ 'testuser2',
+ 'http://osef2.com',
+ ARRAY ['bitcoin', 'etherum'],
+ ARRAY [1, 2]
+);
```
Comment (berthieresteban) [inline backend/scripts/db_sample.sql]: Same

File: backend/routes/users.js [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+  res.status(418).json({
+      error: `bdd request failed`,
+      message: `I'm a teapot`
+  })
+  console.error(e.stack)
+}
+
+/* GET users listing. */
+router.get('/users/login', function (req, res, next) {
+  let query = "SELECT * FROM USERS";
```
Comment (Yoann-Abbes) [inline backend/routes/users.js]: You should name your queries, and should be const

File: backend/app.js [hunk 29]
```diff
@@ -3,9 +3,9 @@ var path = require('path');
 var cookieParser = require('cookie-parser');
```
Comment (berthieresteban) [inline backend/app.js]:
```suggestion
const cookieParser = require('cookie-parser');
```

File: backend/app.js [hunk 30]
```diff
@@ -3,9 +3,9 @@ var path = require('path');
...
 var logger = require('morgan');
```
Comment (berthieresteban) [inline backend/app.js]:
```suggestion
const logger = require('morgan');
```

File: backend/routes/articles.js [hunk 31]
```diff
@@ -2,9 +2,9 @@ var express = require('express');
 var router = express.Router();
```
Comment (berthieresteban) [inline backend/routes/articles.js]:
```suggestion
const router = express.Router();
```

File: backend/routes/users.js [hunk 32]
```diff
@@ -0,0 +1,26 @@
+var express = require('express');
```
Comment (berthieresteban) [inline backend/routes/users.js]:
```suggestion
const express = require('express');
```

File: backend/routes/users.js [hunk 33]
```diff
@@ -0,0 +1,26 @@
...
+var router = express.Router();
```
Comment (berthieresteban) [inline backend/routes/users.js]:
```suggestion
const router = express.Router();
```

File: backend/routes/articles.js [hunk 34]
```diff
@@ -0,0 +1,11 @@
+var express = require('express');
```
Comment (berthieresteban) [inline backend/routes/articles.js]:
```suggestion
const express = require('express');
```

File: backend/app.js [hunk 35]
```diff
@@ -1,11 +1,11 @@
...
 
-var indexRouter = require('./routes/index');
-var userRouter = require('./routes/user');
-var adminRouter = require('./routes/admin');
+const cryptoRouter = require('./routes/cryptos');
+const userRouter = require('./routes/users');
+const articlesRouter = require('./routes/articles');
 const swaggerDoc = require('./swaggerDoc');
 
 var app = express();
```
Comment (berthieresteban) [inline backend/app.js]:
```suggestion
const app = express();
```

File: backend/app.js [hunk 36]
```diff
@@ -1,11 +1,11 @@
 var express = require('express');
```
Comment (berthieresteban) [inline backend/app.js]:
```suggestion
const express = require('express');
```

File: backend/app.js [hunk 37]
```diff
@@ -1,11 +1,11 @@
...
 var path = require('path');
```
Comment (berthieresteban) [inline backend/app.js]:
```suggestion
const path = require('path');
```

File: backend/bin/www [hunk 38]
```diff
@@ -6,22 +6,22 @@ require('dotenv').config()
...
+let app = require('../app');
+let debug = require('debug')('backend:server');
+let http = require('http');
 
 /**
  * Get port from environment and store in Express.
  */
 
-var port = normalizePort(process.env.PORT || '5005');
+let port = normalizePort(process.env.PORT || '5005');
```
Comment (Yoann-Abbes) [inline backend/bin/www]:
```suggestion
const port = normalizePort(process.env.PORT || '5005');
```

File: backend/bin/www [hunk 39]
```diff
@@ -6,22 +6,22 @@ require('dotenv').config()
...
 app.set('port', port);
 
 /**
  * Create HTTP server.
  */
 
-var server = http.createServer(app);
+let server = http.createServer(app);
```
Comment (Yoann-Abbes) [inline backend/bin/www]:
```suggestion
const server = http.createServer(app);
```

**Comment by Gerroll** [comment]
What does this PR do?

This PR add all the routes require in the spec for crypto currency

How should this be manually tested?
You must follow the pull request #10 first because changes were made.
Then please follow the next commands :

    cd backend
    npm i
    npm start
    Try routes define in file backend/routes/cryptos.js (there are some example in it)

Other changes
#10 => field 'period' now match specification by taking the one of the following values : 'daily', 'hourly', 'minute'
/
Boyscout
In the futur, swagger should be implemented
/

**Pull request merged**